service/matching/db_task_manager.go

```diff
@@ -200,72 +217,70 @@ AcquireLoop:
 			break AcquireLoop
 		}
 		if !common.IsPersistenceTransientError(err) {
-			d.Stop()
-			break AcquireLoop
+			d.broadcastStop()
+			return err
 		}
 		time.Sleep(2 * time.Second)
 	}
+	return nil
 }
 
-func (d *dbTaskManager) writerEventLoop(
-	ctx context.Context,
-) {
+func (d *dbTaskManager) writerEventLoop(ctx context.Context) error {
+	defer d.group.Cancel()
 	<-d.startupChan
 	ctx = d.initContext(ctx)
 
-	updateQueueTicker := time.NewTicker(dbTaskUpdateQueueInterval)
+	updateQueueTicker := time.NewTicker(d.dbTaskUpdateQueueInterval)
 	defer updateQueueTicker.Stop()
-	// TODO we should impl a more efficient method to
-	//  buffer & wait for max duration
-	//  right now simply just flush every dbTaskFlushInterval
+	// TODO: Replace buffer logic with #2970
 	flushTicker := time.NewTicker(dbTaskFlushInterval)
 	defer flushTicker.Stop()
 
 	for {
 		if d.isStopped() {
-			return
+			return nil
 		}
 
 		select {
 		case <-d.shutdownChan:
-			return
+			return nil
 		case <-d.taskQueueOwnership.getShutdownChan():
-			d.Stop()
```

Comment:
stop is more preferable (since this function can contain more logic

Reply:
We need to have different behaviors for ongoing read & write, that's why I added goro and introduced a broadcastStop